google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/PublisherImplTest.java

```diff
@@ -1304,6 +1315,70 @@ public void run() {
     publish4Completed.await();
   }
 
+  @Test
+  public void testPublishOpenTelemetryTracing() throws Exception {
+    OpenTelemetryRule openTelemetryTesting = OpenTelemetryRule.create();
+    OpenTelemetry openTelemetry = openTelemetryTesting.getOpenTelemetry();
+    final Publisher publisher =
+        getTestPublisherBuilder()
+            .setBatchingSettings(
+                Publisher.Builder.DEFAULT_BATCHING_SETTINGS
+                    .toBuilder()
+                    .setElementCountThreshold(1L)
+                    .setDelayThreshold(Duration.ofSeconds(5))
+                    .setFlowControlSettings(
+                        FlowControlSettings.newBuilder()
+                            .setLimitExceededBehavior(FlowController.LimitExceededBehavior.Block)
+                            .setMaxOutstandingElementCount(2L)
+                            .setMaxOutstandingRequestBytes(100L)
+                            .build())
+                    .build())
+            .setOpenTelemetry(openTelemetry)
+            .setEnableOpenTelemetryTracing(true)
+            .build();
+
+    testPublisherServiceImpl.addPublishResponse(PublishResponse.newBuilder().addMessageIds("1"));
+    ApiFuture<String> publishFuture = sendTestMessage(publisher, "A");
+    fakeExecutor.advanceTime(Duration.ofSeconds(5));
+    assertEquals("1", publishFuture.get());
+    fakeExecutor.advanceTime(Duration.ofSeconds(5));
+
+    List<SpanData> allSpans = openTelemetryTesting.getSpans();
+    assertEquals(4, allSpans.size());
+    SpanData flowControlSpanData = allSpans.get(0);
+    SpanData batchingSpanData = allSpans.get(1);
+    SpanData publishRpcSpanData = allSpans.get(2);
+    SpanData publisherSpanData = allSpans.get(3);
```

Comment:
How are these orders determined? It doesn't seem like it's chronological. Is it possible these change based on some race condition?

Reply:
The order is decided by when the span is ended. This should be the order in which spans are ended, so there shouldn't be a race condition in which they would end in a different order.